zfit/core/constraint.py

```diff
@@ -0,0 +1,178 @@
+                 name: str = "BaseConstraint", dtype=ztypes.float,
+                 **kwargs):
+        """The base model to inherit from and overwrite `_unnormalized_pdf`.
+
+        Args:
+            dtype (DType): the dtype of the constraint
+            name (str): the name of the constraint
+            params (Dict(str, :py:class:`~zfit.Parameter`)): A dictionary with the internal name of the
+                parameter and the parameters itself the constrains depends on
+        """
+
+        super().__init__(name=name, dtype=dtype, params=params, **kwargs)
+
+    @abc.abstractmethod
+    def _eval(self):
+        raise NotImplementedError
+
+    def value(self):
+        return self._value()
+
+    def _value(self):
+        try:
+            return self._eval()
+
+        except NotImplementedError:
+            raise NotImplementedError("_eval not properly defined!")
+
+    def _get_dependents(self) -> ztyping.DependentsType:
+        return self._extract_dependents(self.get_params())
+
+    def sample(self, n):
+        raise NotImplementedError
+
+
+class SimpleConstraint(BaseConstraint):
+
+    def __init__(self, func: Callable, params: Optional[ztyping.ParametersType] = None):
+        """Constraint from a (function returning a ) Tensor.
+
+        Args:
+            func: Callable that constructs the constraint and returns a tensor.
+            dependents: The dependents (independent `zfit.Parameter`) of the loss. If not given, the
+                dependents are figured out automatically.
+        """
+
+        self._simple_func = func
+        self._simple_func_dependents = convert_to_container(params, container=set)
+        if params is None:
+            params = self.get_dependents()
+
+        params = convert_to_container(params, container=list)
+        params = {p.name: p for p in params}
+
+        super().__init__(name="SimpleConstraint", params=params)
+
+    def _get_dependents(self):
+        dependents = self._simple_func_dependents
+        if dependents is None:
+            independent_params = tf.get_collection("zfit_independent")
+            dependents = get_dependents_auto(tensor=self.value(), candidates=independent_params)
+            self._simple_func_dependents = dependents
+        return dependents
+
+    def _eval(self):
+        return self._simple_func()
+
+
+class ProbConstraint(BaseConstraint):
+
+    def __init__(self, params: Dict[str, ZfitParameter], distribution: Distribution,
+                 name: str = "ProbConstraint", dtype=ztypes.float,
+                 **kwargs):
+        """ Base class for constraints using a probability density function.
+
+        Args:
+            distribution (tensorflow_probability.distributions.Distribution): The probability density function
+                used to constraint the parameters
+
+        """
+
+        super().__init__(params=params, name=name, dtype=dtype, **kwargs)
+
+        self._distribution = distribution
```

Comment:
This is a very tricky one: https://github.com/tensorflow/probability/issues/285
In short, for on-the-fly computations of the normalization (a feature actually  working but not yet extensively tested), we need the `GradientTape`. And this only works, if the distribution is _created while invoking the pdf_.

So what we need is something like in https://github.com/zfit/zfit/blob/ea2f6446a090a55e4572bd3a70ebdc068772a0a1/zfit/models/dist_tfp.py#L56 _for the value_ (for the `sample` it is not actually needed, but let's put it as well).

What is needed in the `__init__` is an argument for the instatiation_params _and_ instantiation args. While I agree that _most probably_ there is no usecase for a changeable parametrization of the Constraint, it still is a cleaner separation.